python_ta/checkers/redundant_assignment_checker.py

```diff
@@ -0,0 +1,130 @@
+    name = 'redundant_assignment'
+    # use dashes for connecting words in message symbol
+    msgs = {'E9959': ('This assignment statement is redundant;' \
+                      ' You can remove it from the program.',
+                      'redundant-assignment',
+                      'This assignment statement is redundant;' \
+                      ' You can remove it from the program.')
+           }
+
+    # this is important so that your checker is executed before others
+    priority = -1
+
+    def __init__(self, linter=None):
+        super().__init__(linter=linter)
+        self._redundant_assignment: Set[astroid.Assign] = set()
+
+    @check_messages('redundant-assignment')
+    def visit_assign(self, node):
+        """Adds message if the assignment statement is <redundant>, i.e. it satisfies
+        the following two properties:
+
+            1. Every path from a definition of a variable `v` to it's first usage
+            goes through at least one re-definition of `v`.
+
+            2. Removing the statement from the program does not in any way change
+            the behavior of the program.
+        """
+        if node in self._redundant_assignment:
+            self.add_message('redundant-assignment', node=node)
+
+    def visit_module(self, node):
+        self._analyze(node)
+
+    def visit_functiondef(self, node):
+        self._analyze(node)
+
+    def _analyze(self, node: Union[astroid.Module, astroid.FunctionDef]) -> None:
+        """Runs the backward data flow algorithm on a `Module` or `Function` CFG, which in turn
+        appends `Assign` nodes to `self.redundant_assignment` if it is redundant.
+
+        Data flow algorithms retrieved from:
+        https://www.seas.harvard.edu/courses/cs252/2011sp/slides/Lec02-Dataflow.pdf#page=31
+        """
+        out_facts = {}
+        cfg = ControlFlowGraph()
+        cfg.start = node.cfg_block
+        worklist = list(cfg.get_blocks_postorder())
+        worklist.reverse()
+
+        all_assigns = self._get_assigns(node)
+        for block in worklist:
+            out_facts[block] = all_assigns.copy()
+
+        while len(worklist) != 0:
+            b = worklist.pop()
+            outs = [out_facts[p.target] for p in b.successors if p.target in out_facts]
+            if outs == []:
+                in_facts = set()
+            else:
+                in_facts = set.intersection(*outs)
+            temp = self._transfer(b, in_facts, node)
+            if temp != out_facts[b]:
+                out_facts[b] = temp
+                worklist.extend([pred.source for pred in b.predecessors])
+
+    def _transfer(self, block: CFGBlock, out_facts: Set[str], scope: Union[astroid.Module, astroid.FunctionDef]) -> Set[str]:
+        gen = out_facts.copy()
+        kill = set()
+        for statement in reversed(block.statements):
+            if isinstance(statement, astroid.Expr) and isinstance(statement.value, astroid.Call) and \
```

Comment:
I don't understand the purpose of this block, please explain.

Reply:
This block specifically addresses function calls of locally defined functions. We're interested in them because they could affect the redundancy of an assignment statement (think `nonlocal` variables). The logic is that if a call to a locally defined function call is made and the local variables of that function are not a subset of `gen`, then we `kill` all the `gen` variables. This is because there is a possibility of usage of one of the `gen` variables within that function definition.

Reply:
Now that I'm thinking about it, the `locals` attribute is not entirely accurate. Should I fix the check? or should remove this entirely so that `gen` variables would be `kill`ed if a call to a locally defined function is made.

Reply:
We've previously discussed locally-defined functions. The bottom line is that I am not interested in tracing behaviour between function calls, at least for now. Please remove this section.